Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36536 Functionalize qconcat and register with new registration API
* #36535 Convert qclamp to new operator registration API.
* #36534 Convert qbatch_norm to new operator registration API.
* #36531 Update quantized README for registering operators with new API.
* #36527 Functionalize qadd and register with new registration API.
* **#36520 Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops.**
* #36513 Add explicit schema for quantized conv/conv_prepack
* #36512 Port detach/detach to new registration API
* #36390 Port Resize to use new registration API.
* #36389 Update reference to RegisterOperators in error message in Convolution
* #36258 Switch to pybind11 style registration function API.

Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

ezyang committed:
Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops.

Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/xnnpack/RegisterOpContextClass.cpp:
```diff
@@ -60,6 +60,15 @@ torch::jit::class_<Conv2dOpContext> register_packed_conv2d_op_context_class() {
 static auto linear_op_context_class = register_packed_linear_op_context_class();
 static auto conv2d_op_context_class = register_packed_conv2d_op_context_class();
 
+// TODO: These should probably be in an xnnpack namespace?
```
The choice of not using xnnpack namespace was explicit as change was made as part of this PR: https://github.com/pytorch/pytorch/pull/35010. Relevant issue: https://github.com/pytorch/pytorch/issues/34913

ezyang replied:
thanks, I'll fix this

kimishpatel reviewed (changes requested):
XNNPACK stuff looks fine to me. Note that namespace change comment I made. So you probably want to remove that TODO.

jerryzh168 commented:
quantized op related changes looks good to me, although we have some in flight PRs but since the changes are small, it should be easy to merge

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

[ghstack-poisoned]

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

[ghstack-poisoned]

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

[ghstack-poisoned]

ezyang committed:
Update on "Add a centralized TORCH_LIBRARY declaration for quantized and xnnpack ops."


Fixes https://github.com/pytorch/pytorch/issues/36510

Note that this introduces some schema string duplication, but I'll
kill the dupes when I migrate the individual files to the new
registration API.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21005369](https://our.internmc.facebook.com/intern/diff/D21005369)

[ghstack-poisoned]

ezyang commented:
This will be rolled up into #36389
